### DISCUSSION TITLE
Writing a plotting library in typs (inspired by matplotlib)

### DISCUSSION THREAD

**Discussion by tmbb** [typst/typst#1444] [Ideas]
Typst seems like an excelent backend for a plotting library, especially if you want to use it from some programming language. One of the best features is the ability to easily measure rendered elements, so that we can position them in inteligent ways. Another interesting feature is the excelent support for math typesetting.

I've hacked together a simple scatterplot with support for axis scales (currently only logarithmic and linear scales, but the scale is just a function so it's easy to add other ones), labels, and other goodies I'm used to from matplotlib (the linked file is self-contained): https://typst.app/project/pJ_qhZdWrFCPnYFIDZx8Ud

My main issue now is how to make this more extensible. For example, I'd like to re-use the "machinery" that plots the axes, ticks, tick labels and all the "decoration" in other plot types besides scatterplots. I'd like to support, for example, line plots, bar plots, histograms, parametric equations (which are actually just line plots). Ideally, I'd like to plot them all in the same plot.

By design, the typst scripting language is quite limited, and doesn't seem very well suited for a project like this (unless I really abuse dictionaries and first class functions, which is actually not that bad). Maybe the best way would be to generate multiple plotting functinos using some kind of template system outside typst?

If anyone is interested, I'd appreciate all help, even if only in conceptualizing the architecture of a reusable library.

**Comment by johannes-wolf** [comment]
@tmbb The Project at https://typst.app/project/pJ_qhZdWrFCPnYFIDZx8Ud is not accessible.

**Comment by tmbb** [comment]
Sorry, that was the private link.
Try this one instead https://typst.app/project/wL4a9kFubyuvg3nCc1olwj

**Comment by tmbb** [comment]
In the linked project I've implemented the beginning of a "low level" backend which assumes that data coordinates have been transformed into what floats between 0 and 1 which represent coordinates relative to the axes' size. Axes support padding (whitespace at the ends) and margins (the parts of the axes before and after the "axis space" starts). The plots support an unlimited number of axes in any of the four sides (top, left, bottom, right).

Currently, the actual type of plots one can use is scatterplots (and a lot is still hardcoded). This doesn't bother me too much because plotting the points is usually the easy part. I'm quite happy to have solved some of the harder parts, such as placing, sizing and stacking the axes.

The `measure(object, styles)` is amazing, and allows the author of the document to imeplement quite complex layouts in "user space" with no support from the lower level layers.
